crates/milli/src/fields_ids_map/metadata.rs

```diff
@@ -106,76 +121,247 @@ impl Metadata {
+    ///
+    /// This is used for testing, prefer using `MetadataBuilder::from_index` instead.
     pub fn new(
-        searchable_attributes: Vec<String>,
-        filterable_attributes: HashSet<String>,
+        searchable_attributes: Option<Vec<String>>,
+        filterable_attributes: Vec<FilterableAttributesRule>,
+        sortable_attributes: HashSet<String>,
+        localized_attributes: Option<Vec<LocalizedAttributesRule>>,
+        distinct_attribute: Option<String>,
+        asc_desc_attributes: HashSet<String>,
+    ) -> Self {
+        Self::_new(
+            searchable_attributes,
+            filterable_attributes,
+            sortable_attributes,
+            localized_attributes,
+            distinct_attribute,
+            asc_desc_attributes,
+        )
+    }
+
+    fn _new(
+        searchable_attributes: Option<Vec<String>>,
+        filterable_attributes: Vec<FilterableAttributesRule>,
         sortable_attributes: HashSet<String>,
         localized_attributes: Option<Vec<LocalizedAttributesRule>>,
+        distinct_attribute: Option<String>,
+        asc_desc_attributes: HashSet<String>,
     ) -> Self {
+        let searchable_attributes = match searchable_attributes {
+            Some(fields) if fields.iter().any(|f| f == "*") => None,
+            Some(fields) => Some(fields),
+            None => None,
+        };
+
         Self {
             searchable_attributes,
             filterable_attributes,
             sortable_attributes,
             localized_attributes,
+            distinct_attribute,
+            asc_desc_attributes,
         }
     }
 
     pub fn metadata_for_field(&self, field: &str) -> Metadata {
-        let searchable = self
-            .searchable_attributes
+        if is_faceted_by(field, RESERVED_VECTORS_FIELD_NAME) {
+            // Vectors fields are not searchable, filterable, distinct or asc_desc
+            return Metadata {
+                searchable: None,
+                sortable: false,
+                distinct: false,
+                asc_desc: false,
+                geo: false,
+                localized_attributes_rule_id: None,
+                filterable_attributes_rule_id: None,
+            };
+        }
+
+        // A field is sortable if it is faceted by a sortable attribute
+        let sortable = self
+            .sortable_attributes
             .iter()
-            .any(|attribute| attribute == "*" || attribute == field);
+            .any(|pattern| match_field_legacy(pattern, field) == PatternMatch::Match);
 
-        let filterable = self.filterable_attributes.contains(field);
+        let filterable_attributes_rule_id = self
+            .filterable_attributes
+            .iter()
+            .position(|attribute| attribute.match_str(field) == PatternMatch::Match)
+            // saturating_add(1): make `id` `NonZero`
+            .map(|id| NonZeroU16::new(id.saturating_add(1).try_into().unwrap()).unwrap());
```

Comment:
Do we enforce anywhere in the settings that the max number of rules is `NonZeroU16::MAX - 1`? If not, should we?

Reply:
No but it's the same for the localized attribute rules 🤔

Reply:
Created as https://github.com/meilisearch/meilisearch/issues/5409, resolving for now